sw/device/lib/testing/mock_mmio_test.cc

```diff
@@ -0,0 +1,32 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/testing/mock_mmio.h"
+
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/mmio.h"
+
+namespace {
+/**
+ * Exercises the register |dev| by reading a value at offset 0x0,
+ * writing it and its completement to 0x4 in sequence, and
+ * returning it.
+ */
+uint32_t WriteTwice(mmio_region_t dev) {
+  auto value = mmio_region_read32(dev, 0x0);
+  mmio_region_write32(dev, 0x4, value);
+  mmio_region_write32(dev, 0x4, ~value);
+  return value;
+}
+
+class WriteTwiceTest : public testing::Test, public mock_mmio::MmioTest {};
+
+TEST_F(WriteTwiceTest, WriteTwice) {
+  EXPECT_READ32(0x0, 0xdeadbeef);
```

Comment:
Just out of curiosity, does gmock initialise values to 0xdeadbeef?

Reply:
No. It's just a random value I picked. EXPECT_READ has the semantics of saying "I expect a read to this offset, and that read will produce this value."

Reply:
I am still confused how in this line of code value of variable `value` is `0xdeadbeef`:
` auto value = mmio_region_read32(dev, 0x0);`

I have grepped for `0xdeadbeef`, but couldn't find anywhere else it is mentioned. That suggests that `EXPECT_READ32(0x0, 0xdeadbeef)` sets the value of "address" `0x0` to `0xdeadbeef`. If this is what is happening and I am not missing something, it's pretty confusing.

Reply:
Okay, I am no longer confused. However I think it would be useful to expand `EXPECT_READn` macro descriptions to something like:
```
/**                                                                                                                                                                                                 
 * Expect a read to the device |dev| at the given offset, returning the given                                                                                                                       
 * 8-bit value. Call to this macro specifies what the next 8 bit read in the code
 * under test will return.
....
```

And/or heavily comment every line in this test to something like:
```
// Specify the value that will be read by the call to `mmio_region_read32`
// in WriteTwice.
EXPECT_READ32(0x0, 0xdeadbeef);
```

It is a very small test, so couple extra comments would not take much time. I think this test being essentially an example of usage, there is no harm to be more explicit than you would normally be.